src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+            cache: Cache::new(config.max_cache_entries),
+            config,
+            clock,
+        }
+    }
+}
+
+impl<Upstream> Debug for Connection<Upstream> {
+    fn fmt(&self, _: &mut Formatter<'_>) -> Result<(), core::fmt::Error> {
+        todo!()
+    }
+}
+
+//------------ SendRequest ----------------------------------------------------
+
+impl<CR, Upstream, C> SendRequest<CR> for Connection<Upstream, C>
+where
+    CR: Clone + ComposeRequest + 'static,
+    Upstream: Clone + SendRequest<CR> + Send + Sync + 'static,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    fn send_request(
+        &self,
+        request_msg: CR,
+    ) -> Box<dyn GetResponse + Send + Sync> {
+        Box::new(Request::<CR, Upstream, C>::new(
+            request_msg,
+            self.upstream.clone(),
+            self.cache.clone(),
+            self.config.clone(),
+            self.clock.clone(),
+        ))
+    }
+}
+
+//------------ Request --------------------------------------------------------
+
+/// The state of a request that is executed.
+pub struct Request<CR, Upstream, C>
+where
+    CR: Send + Sync,
+    Upstream: Send + Sync,
+    C: Clock + Send + Sync,
+{
+    /// State of the request.
+    state: RequestState,
+
+    /// The request message.
+    request_msg: CR,
+
+    /// The upstream transport of the connection.
+    upstream: Upstream,
+
+    /// The cache of the connection.
+    cache: Cache<Key, Arc<Value<C>>>,
+
+    /// The configuration of the connection.
+    config: Config,
+
+    /// The clock to use for expiring cache entries.
+    clock: C,
+}
+
+impl<CR, Upstream, C> Request<CR, Upstream, C>
+where
+    CR: Clone + ComposeRequest + Send + Sync,
+    Upstream: SendRequest<CR> + Send + Sync,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    /// Create a new Request object.
+    fn new(
+        request_msg: CR,
+        upstream: Upstream,
+        cache: Cache<Key, Arc<Value<C>>>,
+        config: Config,
+        clock: C,
+    ) -> Request<CR, Upstream, C> {
+        Self {
+            state: RequestState::Init,
+            request_msg,
+            upstream,
+            cache,
+            config,
+            clock,
+        }
+    }
+
+    /// This is the implementation of the get_response method.
+    // This function is cancel safe.
+    async fn get_response_impl(&mut self) -> Result<Message<Bytes>, Error> {
+        loop {
+            match &mut self.state {
+                RequestState::Init => {
+                    let msg = self.request_msg.to_message()?;
+                    let header = msg.header();
+                    let opcode = header.opcode();
+
+                    // Extract Qname, Qclass, Qtype
+                    let mut question_section = msg.question();
+                    let question = match question_section.next() {
+                        None => {
+                            // No question. Just forward the request.
+                            let request = self
```

Comment:
If I were writing this I would factor out the common code in all the `GetResponseNoCache` then `continue` blocks, but, even if you agree, I don't insist on it being done right now.